docs/design/hyperNode-auto-discovery.md

```diff
@@ -0,0 +1,196 @@
+    *   Send network topology information to `HyperNode Controller`.
+*   **Key Code**:
+
+```go
+// RegisterDiscoverer registers the discoverer constructor for the specified source
+func RegisterDiscoverer(source string, constructor DiscovererConstructor) {
+    discovererRegistry[source] = constructor
+}
+```
+
+### Discoverer
+
+`Discoverer` is a specific network topology discoverer, responsible for obtaining network topology information from a specific data source and converting it into `HyperNode` resources.
+
+*   **Function**:
+    *   Obtain network topology information from the specified data source.
+    *   Convert network topology information into `HyperNode` resources.
+*   **Interface Definition**:
+
+```go
+// Discoverer is the interface for network topology discovery
+type Discoverer interface {
+    // Start begins the discovery process, sending discovered nodes through the provided channel
+    Start(outputCh chan<- []*topologyv1alpha1.HyperNode) error
+
+    // Stop halts the discovery process
+    Stop() error
+
+    // Name returns the discoverer identifier, this is used for labeling discovered hyperNodes for distinction.
+    Name() string
+}
+```
+
+*   **Implementation**:
+    *   **UFM Discoverer**: Obtain network topology information from UFM (Unified Fabric Manager).
+    *   **RoCE Discoverer**: Obtain network topology information from RoCE (RDMA over Converged Ethernet) devices.
+    *   **Label Discoverer**: Discover network topology information based on the Label on the Node.
+
+### HyperNode Controller
+
+`HyperNode Controller` is responsible for listening to changes in `HyperNode` resources and creating, updating, or deleting `HyperNode` resources based on the discovered topology information.
+
+*   **Function**:
+    *   Listen to change events of `HyperNode` resources.
+    *   Create, update, or delete `HyperNode` resources based on the network topology information provided by `Discoverer`.
+*   **Key Code**:
+
+```go
+func (hn *hyperNodeController) reconcileTopology(source string, discoveredNodes []*topologyv1alpha1.HyperNode) {
+ klog.InfoS("Starting topology reconciliation", "source", source, "discoveredNodeCount", len(discoveredNodes))
+
+ existingNodes, err := hn.hyperNodeLister.List(labels.SelectorFromSet(labels.Set{
+  api.NetworkTopologySourceLabelKey: source,
+ }))
+ if err != nil {
+  klog.ErrorS(err, "Failed to list existing HyperNode resources")
+  return
+ }
+
+ existingNodeMap := make(map[string]*topologyv1alpha1.HyperNode)
+ for _, node := range existingNodes {
+  existingNodeMap[node.Name] = node
+ }
+
+ discoveredNodeMap := make(map[string]*topologyv1alpha1.HyperNode)
+ for _, node := range discoveredNodes {
+  if node.Labels == nil {
+   node.Labels = make(map[string]string)
+  }
+  node.Labels[api.NetworkTopologySourceLabelKey] = source
+  discoveredNodeMap[node.Name] = node
+ }
+
+ for name, node := range discoveredNodeMap {
+  if _, exists := existingNodeMap[name]; !exists {
+   [...]
```

Comment:
Why is the create and update logic hidden here, but the delete logic is displayed in L183-L188? 😂